codec/src/main/java/io/netty/handler/codec/compression/ZstdDecoder.java

```diff
@@ -0,0 +1,184 @@
+import io.netty.buffer.ByteBufOutputStream;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.TooLongFrameException;
+import io.netty.util.internal.ObjectUtil;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import static io.netty.handler.codec.compression.ZstdConstants.DEFAULT_MAX_BLOCK_SIZE;
+
+/**
+ * Decompresses a compressed block {@link ByteBuf} using the Zstandard algorithm.
+ * See <a href="https://facebook.github.io/zstd">Zstandard</a>.
+ */
+public final class ZstdDecoder extends ByteToMessageDecoder {
+    private static final int COPY_BUF_SIZE = 8024;
+    private final int maxBlockSize;
+    private final MutableByteBufInputStream inputStream = new MutableByteBufInputStream();
+    private ZstdInputStreamNoFinalizer zstdIs;
+    private ByteBufOutputStream outputStream;
+
+    private volatile State currentState = State.DECOMPRESS_DATA;
+
+    /**
+     * Creates a new Zstd decoder.
+     *
+     * Please note that if you use the default constructor, the MAX_BLOCK_SIZE
+     * will be used. If you want to specify MAX_BLOCK_SIZE yourself,
+     * please use {@link ZstdDecoder(int)} constructor
+     */
+    public ZstdDecoder() {
+        this(DEFAULT_MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd decoder.
+     *  @param  maxBlockSize
+     *            specifies the max block size
+     */
+    public ZstdDecoder(int maxBlockSize) {
+        this.maxBlockSize = ObjectUtil.checkPositive(maxBlockSize, "maxBlockSize");
+    }
+
+    /**
+     * Current state of stream.
+     */
+    private enum State {
+        DECOMPRESS_DATA,
+        CORRUPTED
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (currentState == State.CORRUPTED) {
+            in.skipBytes(in.readableBytes());
+            return;
+        }
+        try {
+            final int compressedLength = in.readableBytes();
+            if (compressedLength > maxBlockSize) {
+                in.skipBytes(compressedLength);
+                throw new TooLongFrameException("too large message: " + compressedLength + " bytes");
+            }
+
+            inputStream.current = in;
+
+            // Let's start with the compressedLength * 2 as often we will not have everything
+            // we need in the in buffer and don't want to reserve too much memory.
+            outputStream = new ByteBufOutputStream(ctx.alloc().buffer(compressedLength * 2));
+
+            copyStream(zstdIs, outputStream);
+            int decompressedLength = outputStream.buffer().readableBytes();
+            if (decompressedLength == 0) {
+                // Did not decompress anything, let's close the output and release the buffer.
+                closeOutputSilently(true);
+                return;
+            }
```

Comment:
What if the input buffer does not have one complete Zstd block?

Reply:
Fixed it